lib/src/rules/parameter_assignments.dart

```diff
@@ -124,75 +106,95 @@ class ParameterAssignments extends LintRule {
-  void visitMethodDeclaration(MethodDeclaration node) {
-    var parameterList = node.parameters;
-    if (parameterList != null) {
-      // Getters don't have parameters.
-      for (var e in parameterList.parameters) {
-        var declaredElement = e.declaredElement;
-        if (declaredElement != null &&
-            node.body.isPotentiallyMutatedInScope(declaredElement)) {
-          _reportIfSimpleParameterOrWithDefaultValue(e, node);
-        }
+  visitPostfixExpression(PostfixExpression node) {
+    if (paramIsNotNullByDefault) {
+      var operand = node.operand;
+      if (operand is SimpleIdentifier &&
+          operand.staticElement == parameter.declaredElement) {
+        rule.reportLint(node);
       }
     }
+
+    super.visitPostfixExpression(node);
   }
 
-  void _reportIfSimpleParameterOrWithDefaultValue(
-      FormalParameter parameter, AstNode functionOrMethodDeclaration) {
-    var nodes = functionOrMethodDeclaration.traverseNodesInDFS();
-
-    if (parameter is SimpleFormalParameter ||
-        _isDefaultFormalParameterWithDefaultValue(parameter)) {
-      var mutatedNodes = nodes.where((n) =>
-          (n is AssignmentExpression &&
-              _isFormalParameterReassigned(parameter, n)) ||
-          _preOrPostFixExpressionMutation(parameter, n));
-      mutatedNodes.forEach(rule.reportLint);
-      return;
+  @override
+  visitPrefixExpression(PrefixExpression node) {
+    if (paramIsNotNullByDefault) {
+      var operand = node.operand;
+      if (operand is SimpleIdentifier &&
+          operand.staticElement == parameter.declaredElement) {
+        rule.reportLint(node);
+      }
     }
 
-    var assignmentsNodes = nodes
-        .where((n) =>
-            n is AssignmentExpression &&
-            _isDefaultFormalParameterWithoutDefaultValueReassigned(
-                parameter, n))
-        .toList();
-
-    var nonNullCoalescingAssignments = assignmentsNodes.where((n) =>
-        (n as AssignmentExpression).operator.type !=
-        TokenType.QUESTION_QUESTION_EQ);
-
-    if (assignmentsNodes.length > 1 ||
-        nonNullCoalescingAssignments.isNotEmpty) {
-      var node = assignmentsNodes.length > 1
-          ? assignmentsNodes.last
-          : nonNullCoalescingAssignments.isNotEmpty
-              ? nonNullCoalescingAssignments.first
-              : parameter;
-      rule.reportLint(node);
+    super.visitPrefixExpression(node);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  final LintRule rule;
+
+  _Visitor(this.rule);
+
+  @override
+  void visitFunctionDeclaration(FunctionDeclaration node) {
+    var parameterList = node.functionExpression.parameters;
+    if (parameterList == null) return;
+    for (var e in parameterList.parameters) {
+      var declaredElement = e.declaredElement;
+      if (declaredElement != null &&
+          node.functionExpression.body
+              .isPotentiallyMutatedInScope(declaredElement)) {
+        node.accept(_DeclarationVisitor(e, rule));
```

Comment:
We only need to visit the body, both here and below. Probably not a measurable improvement in our measurement code base, but it means that we could unify these two visit methods so that they looked something like
```
void visitFunctionDeclaration(FunctionDeclaration node) {
  _checkParameters(
    node.functionExpression.parameters,
    node.functionExpression.body,
  );
}
```

Reply:
Good call.  Thanks!